src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -179,7 +179,7 @@ private Task<Releaser> LockWaitingHelper(Task<bool> waitTask)
         {
             Requires.NotNull(waitTask, nameof(waitTask));
 
-            return waitTask.IsCompleted
+            return waitTask.Status == TaskStatus.RanToCompletion
```

Comment:
This change isn't necessary. We already call `waitTask.Result` which will throw instead of returning the `uncontestedReleaser` so it's unrelated to this bug.

Reply:
Reverted this one. It's still a bug, but not as severe and can be implemented separately with its own test.